File: Apps/iOS/Features/Harmonica/HarmonicaViewModel.swift
import Combine
import CoreGraphics
import HarmonicaCore

@MainActor
final class HarmonicaViewModel: ObservableObject {
    @Published private(set) var state: HarmonicaViewState = .preparingSound

    private let playHarmonica: PlayHarmonicaUseCase
    private let playScore: PlayScoreUseCase
    private let tilt: TiltProtocol
    private let settingsRepository: SettingsRepository
    private let tunes: [Score]
    private let presenter: HarmonicaPresenter
    private let log: LogProtocol
    private let pauseBeforeATune: Duration
    private let openSettings: @MainActor () -> Void
    private var settings: PlayerSettings
    private var isOnScreen = false
    private var soundPreparation: Task<Void, Never>?
    private var tiltFollowing: Task<Void, Never>?
    private var tunePerformance: Task<Void, Never>?
    private var playingTune: Int?
    private var settingsSubscription: AnyCancellable?

    // MARK: - Public

    init(
        playHarmonica: PlayHarmonicaUseCase,
        playScore: PlayScoreUseCase,
        tilt: TiltProtocol,
        settingsRepository: SettingsRepository,
        tunes: [Score],
        presenter: HarmonicaPresenter,
        log: LogProtocol,
        pauseBeforeATune: Duration,
        openSettings: @escaping @MainActor () -> Void
    ) {
        self.playHarmonica = playHarmonica
        self.playScore = playScore
        self.tilt = tilt
        self.settingsRepository = settingsRepository
        self.tunes = tunes
        self.presenter = presenter
        self.log = log
        self.pauseBeforeATune = pauseBeforeATune
        self.openSettings = openSettings
        settings = settingsRepository.settings()
        settingsSubscription = settingsRepository.savedSettings.sink { [weak self] savedSettings in self?.adopt(savedSettings) }
    }

    deinit {
        soundPreparation?.cancel()
        tiltFollowing?.cancel()
        tunePerformance?.cancel()
    }

    func send(_ action: HarmonicaAction) {
        switch action {
        case .appBecameActive:
            prepareSound()
        case .appLeftTheForeground:
            silence()
        case .screenAppeared:
            isOnScreen = true
            followTheTiltIfItShould()
        case .screenDisappeared:
            isOnScreen = false
            silence()
            followTheTiltIfItShould()
        case .stripTouched(let touches, across: let size):
            play(touches, across: size)
        case .keySliderMoved(toPosition: let position):
            changeKey(toPosition: position)
        case .tuneChosen(at: let index):
            playTheTune(at: index)
        case .stopTuneButtonTapped:
            stopTheTune()
        case .shapingPadTouched(let touches, across: let size):
            shapeTone(with: touches, across: size)
        case .shapingPadPinched(by: let magnification, within: let area):
            resizeShapingPad(by: magnification, within: area)
        case .settingsButtonTapped:
            openSettings()
        }
    }

    func marksOnTheStrip(
        for touches: [FingerTouch],
        across size: CGSize,
        style: HarmonicaViewState.FingerMarks
    ) -> [FingerMark] {
        StripTouchMapper(touches, across: size).marks(drawn: style)
    }

    func marksOnTheShapingPad(for touches: [FingerTouch], across size: CGSize) -> [FingerMark] {
        ShapingPadTouchMapper(touches, across: size).marks
    }

    // MARK: - Private

    private var reasonNotToFollowTheTilt: String? {
        guard case .ready = state else { return "the sound is not ready" }
        guard isOnScreen else { return "the harmonica is not on screen" }
        guard settings.isCuppingEnabled else { return "cupping is off in the settings" }

        return nil
    }

    private func prepareSound() {
        soundPreparation?.cancel()
        soundPreparation = Task { [weak self] in await self?.prepareTheEngine() }
    }

    private func play(_ touches: [FingerTouch], across size: CGSize) {
        guard isReady(toTake: "a touch on the strip") else { return }

        if !touches.isEmpty {
            stopTheScore()
        }
        show(playHarmonica.play(at: StripTouchMapper(touches, across: size).positions))
    }

    private func changeKey(toPosition position: Double) {
        guard isReady(toTake: "a key change") else { return }

        let key = HarmonicaKey(nearestSliderPosition: Int(position.rounded()))
        show(playHarmonica.changeKey(to: key))
    }

    private func playTheTune(at index: Int) {
        guard isReady(toTake: "a tune") else { return }
        guard tunes.indices.contains(index) else {
            assertionFailure("the menu offered tune \(index) of \(tunes.count)")
            log.record("tune \(index) was asked for, there are \(tunes.count), nothing plays")
            return
        }

        stopTheScore()
        playingTune = index
        tunePerformance = perform(tunes[index])
        show(playHarmonica.harmonica)
    }

    private func stopTheTune() {
        guard isReady(toTake: "stopping the tune") else { return }

        stopTheScore()
        show(playHarmonica.stopPlaying(.ringsDown))
    }

    private func shapeTone(with touches: [FingerTouch], across size: CGSize) {
        let shapingPad = ShapingPadTouchMapper(touches, across: size)
        guard let shaping = shapingPad.pitchShaping, let vibrato = shapingPad.vibrato else {
            stopShapingTone()
            return
        }

        shapeTone(shaping, vibrato: vibrato)
    }

    private func resizeShapingPad(by magnification: CGFloat, within area: CGSize) {
        guard isReady(toTake: "a pinch on the shaping pad") else { return }

        var settingsAfterThePinch = settings
        settingsAfterThePinch.shapingPadSize = ShapingPadSizing(in: area)
            .size(afterPinching: settings.shapingPadSize, by: magnification)
        guard settingsAfterThePinch != settings else { return }

        log.recordSample("shaping pad resized to \(settingsAfterThePinch.shapingPadSize.fraction) of its range")
        settingsRepository.save(settingsAfterThePinch)
    }

    private func stopShapingTone() {
        shapeTone(.rest, vibrato: .off)
    }

    private func prepareTheEngine() async {
        do throws(AudioEngineError) {
            _ = try await playHarmonica.prepare()
            show(harmonica(adopting: settings))
        } catch {
            publish(presenter.presentSoundUnavailable(because: error))
        }
        followTheTiltIfItShould()
    }

    private func adopt(_ savedSettings: PlayerSettings) {
        settings = savedSettings
        followTheTiltIfItShould()
        guard isReady(toTake: "the settings") else { return }

        show(harmonica(adopting: savedSettings))
    }

    private func harmonica(adopting settings: PlayerSettings) -> Harmonica {
        let harmonicaInTheChosenStyle = playHarmonica.changeStyle(to: settings.style)
        guard !settings.isCuppingEnabled else { return harmonicaInTheChosenStyle }

        return playHarmonica.cupHands(to: .open)
    }

    private func followTheTiltIfItShould() {
        guard let reason = reasonNotToFollowTheTilt else {
            startFollowingTheTilt()
            return
        }

        stopFollowingTheTilt(because: reason)
    }

    private func startFollowingTheTilt() {
        guard tiltFollowing == nil else { return }

        log.record("following the lean of the phone")
        let leaningsOfThePhone = tilt.tiltToTheRight()
        tiltFollowing = Task { [weak self] in
            for await leaning in leaningsOfThePhone {
                self?.cupHands(to: CupDepth(clamping: leaning))
            }
        }
    }

    private func stopFollowingTheTilt(because reason: String) {
        guard let tiltFollowing else { return }

        tiltFollowing.cancel()
        self.tiltFollowing = nil
        log.record("the lean of the phone is no longer followed: \(reason)")
    }

    private func silence() {
        guard isReady(toTake: "silencing the harmonica") else { return }

        stopTheScore()
        show(playHarmonica.stopPlaying(.ringsDown))
        stopShapingTone()
    }

    private func perform(_ tune: Score) -> Task<Void, Never> {
        Task { [weak self, playScore, pauseBeforeATune] in
            do {
                try await Task.sleep(for: pauseBeforeATune)
            } catch {
                return
            }

            await playScore.play(tune) { harmonica in self?.show(harmonica) }
            guard !Task.isCancelled else { return }

            self?.finishTheTune()
        }
    }

    private func finishTheTune() {
        tunePerformance = nil
        playingTune = nil
        show(playHarmonica.stopPlaying(.ringsDown))
    }

    private func shapeTone(_ shaping: PitchShaping, vibrato: VibratoDepth) {
        guard isReady(toTake: "a touch on the shaping pad") else { return }

        show(playHarmonica.shapeTone(shaping, vibrato: vibrato))
    }

    private func cupHands(to cup: CupDepth) {
        guard isReady(toTake: "a lean of the phone") else { return }

        show(playHarmonica.cupHands(to: cup))
    }

    private func isReady(toTake action: String) -> Bool {
        guard case .ready = state else {
            log.recordSample("\(action) arrived while the sound was not ready, nothing changed")
            return false
        }

        return true
    }

    private func stopTheScore() {
        tunePerformance?.cancel()
        tunePerformance = nil
        playingTune = nil
    }

    private func show(_ harmonica: Harmonica) {
        publish(presenter.present(harmonica, settings: settings, playingTune: playingTune))
    }

    private func publish(_ updated: HarmonicaViewState) {
        guard updated != state else { return }

        state = updated
    }
}

File: Web/src/features/harmonica/harmonicaViewModel.ts
import type { HarmonicaDTO } from '../../core/harmonicaDTO.js'
import { holesWideOfTheContact } from '../../core/harmonicaDTO.js'
import type { HarmonicaCore } from '../../core/harmonicaCore.js'
import type { AudioEngine, AudioEngineError } from '../../core/audioEngine.js'
import { AudioEngineFailure } from '../../core/audioEngine.js'
import type { Log } from '../../logging/log.js'
import type { HarmonicaAction } from './harmonicaAction.js'
import type { HarmonicaPresenter } from './harmonicaPresenter.js'
import type { AreaSize } from './touch/areaSize.js'
import type { Contact } from './touch/contact.js'
import type { FingerMark } from './touch/fingerMark.js'
import type { FingerMarksViewState, HarmonicaViewState, PlayingStyleChoice } from './harmonicaViewState.js'
import { isPlayingStyleChoice } from './harmonicaViewState.js'
import { ShapingPadTouchMapper } from './touch/shapingPadTouchMapper.js'
import { StripTouchMapper } from './touch/stripTouchMapper.js'
import type { TuneEnding, TunePerformance } from './playTheTuneUseCase.js'
import { PlayTheTuneUseCase } from './playTheTuneUseCase.js'

const startingNotesPerFinger = 1
const smallestShapingPadScale = 0.45
const largestShapingPadScale = 2

export class HarmonicaViewModel {
    private state: HarmonicaViewState = { kind: 'preparingSound' }
    private performance: TunePerformance | null = null
    private notesPerFinger = startingNotesPerFinger
    private shapingPadScale = 1
    private shownHarmonica: HarmonicaDTO | null = null
    private show: (state: HarmonicaViewState) => void = () => {}

    constructor(
        private readonly core: HarmonicaCore,
        private readonly audio: AudioEngine,
        private readonly tunePlayer: PlayTheTuneUseCase,
        private readonly presenter: HarmonicaPresenter,
        private readonly log: Log
    ) {}

    onStateChanged(show: (state: HarmonicaViewState) => void): void {
        this.show = show
        show(this.state)
    }

    send(action: HarmonicaAction): void {
        switch (action.kind) {
            case 'startButtonTapped': return void this.prepareSound()
            case 'pageHidden': return this.silence()
            case 'stripTouched': return this.play(action.contacts, action.across)
            case 'keySliderMoved': return this.changeKey(action.position)
            case 'styleChosen': return this.changeStyle(action.named)
            case 'notesPerFingerChosen': return this.changeNotesPerFinger(action.notes)
            case 'tuneChosen': return this.playTheTune(action.index)
            case 'stopTuneButtonTapped': return this.stopTheTuneAndSilence()
            case 'shapingPadTouched': return this.shapeTone(action.contacts, action.across)
            case 'shapingPadPinched': return this.resizeShapingPad(action.magnification)
        }
    }

    marksOnTheStrip(contacts: readonly Contact[], across: AreaSize, drawn: FingerMarksViewState): FingerMark[] {
        return new StripTouchMapper(contacts, across).marks(drawn)
    }

    marksOnTheShapingPad(contacts: readonly Contact[], across: AreaSize): FingerMark[] {
        return new ShapingPadTouchMapper(contacts, across).marks()
    }

    private async prepareSound(): Promise<void> {
        try {
            await this.audio.prepare()
        } catch (failure) {
            this.publish(this.presenter.presentSoundUnavailable(this.engineErrorOf(failure)))
            return
        }
        this.core.start()
        this.present(this.core.changeMouth(this.notesPerFinger))
    }

    private play(contacts: readonly Contact[], across: AreaSize): void {
        if (!this.isReady('a touch on the strip')) return

        if (contacts.length > 0) this.stopTheTune()
        this.present(this.core.playAt(new StripTouchMapper(contacts, across).positions()))
    }

    private changeKey(toPosition: number): void {
        if (!this.isReady('a key change')) return

        this.present(this.core.changeKey(Math.round(toPosition)))
    }

    private changeStyle(named: string): void {
        if (!this.isReady('a playing style')) return
        if (!isPlayingStyleChoice(named)) {
            console.assert(false, `the menu offered a playing style called ${named}`)
            this.log.record(`playing style ${named} was asked for and the page does not have it, nothing changed`)
            return
        }

        this.core.changeStyle(styleIndex(named))
        this.present(this.core.changeMouth(this.holesWideFor(named)))
    }

    private changeNotesPerFinger(notes: number): void {
        if (!this.isReady('a number of notes per finger')) return

        this.notesPerFinger = notes
        this.present(this.core.changeMouth(notes))
    }

    private playTheTune(index: number): void {
        if (!this.isReady('a tune')) return
        if (!this.offersTheTune(index)) {
            console.assert(false, `the menu offered tune ${index}, which the page does not have`)
            this.log.record(`tune ${index} was asked for and the page does not have it, nothing plays`)
            return
        }

        this.stopTheTune()
        this.log.record(`the tune ${this.nameOfTune(index)} started`)
        this.performance = this.tunePlayer.play(index, harmonica => this.present(harmonica))
        void this.performance.finished
            .then(ending => this.tuneEnded(ending))
            .catch(failure => this.tuneFailed(failure))
    }

    private stopTheTuneAndSilence(): void {
        if (!this.isReady('stopping the tune')) return

        this.stopTheTune()
        this.present(this.core.stopPlaying(true))
    }

    private shapeTone(contacts: readonly Contact[], across: AreaSize): void {
        const shaping = new ShapingPadTouchMapper(contacts, across).shaping()
        if (shaping === null) return this.stopShapingTone()

        this.shapeToneTo(shaping.pitch, shaping.vibrato)
    }

    private resizeShapingPad(magnification: number): void {
        if (!this.isReady('a pinch on the shaping pad') || this.shownHarmonica === null) return

        this.shapingPadScale = Math.min(
            largestShapingPadScale,
            Math.max(smallestShapingPadScale, this.shapingPadScale * magnification)
        )
        this.log.recordSample(`shaping pad scaled to ${this.shapingPadScale.toFixed(2)} of its natural side`)
        this.present(this.shownHarmonica)
    }

    private silence(): void {
        if (!this.isReady('silencing the harmonica')) return

        this.stopTheTune()
        this.present(this.core.stopPlaying(true))
        this.stopShapingTone()
    }

    private stopShapingTone(): void {
        this.shapeToneTo(0, 0)
    }

    private shapeToneTo(pitch: number, vibrato: number): void {
        if (!this.isReady('a touch on the shaping pad')) return

        this.present(this.core.shapeTone(pitch, vibrato))
    }

    private stopTheTune(): void {
        this.performance?.cancel()
        this.performance = null
    }

    private holesWideFor(style: PlayingStyleChoice): number {
        return style === 'oneFingerNotesByPressure' ? holesWideOfTheContact : this.notesPerFinger
    }

    private isReady(action: string): boolean {
        if (this.state.kind === 'ready') return true

        this.log.recordSample(`${action} arrived while the sound was ${this.state.kind}, nothing changed`)
        return false
    }

    private offersTheTune(index: number): boolean {
        return this.state.kind === 'ready' && this.state.playable.demo.tunes.some(tune => tune.id === index)
    }

    private engineErrorOf(failure: unknown): AudioEngineError {
        if (failure instanceof AudioEngineFailure) return failure.error

        console.assert(false, `the audio engine rejected with ${describe(failure)}, which it never should`)
        this.log.record(`the audio engine rejected with ${describe(failure)}, shown as an output that would not start`)
        return { kind: 'outputRefused' }
    }

    private nameOfTune(index: number): string {
        if (this.state.kind !== 'ready') return String(index)

        return this.state.playable.demo.tunes.find(tune => tune.id === index)?.name ?? String(index)
    }

    private tuneEnded(ending: TuneEnding): void {
        if (ending === 'stopped') return this.log.record('the tune was stopped')

        this.log.record('the tune ended')
        this.silenceTheTune()
    }

    private tuneFailed(failure: unknown): void {
        this.log.record(`the tune stopped: ${describe(failure)}`)
        this.silenceTheTune()
    }

    private silenceTheTune(): void {
        this.performance = null
        this.present(this.core.stopPlaying(true))
    }

    private present(harmonica: HarmonicaDTO): void {
        this.shownHarmonica = harmonica
        this.publish(this.presenter.present(harmonica, this.performance !== null, this.shapingPadScale))
    }

    private publish(state: HarmonicaViewState): void {
        this.state = state
        this.show(state)
    }
}

function styleIndex(choice: PlayingStyleChoice): number {
    switch (choice) {
        case 'severalFingersSeveralNotes': return 0
        case 'severalFingersOneNote': return 1
        case 'oneFingerSeveralNotes': return 2
        case 'oneFingerNotesByPressure': return 2
    }
}

function describe(failure: unknown): string {
    return failure instanceof Error ? `${failure.name}: ${failure.message}` : String(failure)
}
